fabric-model-loading-api-v1/src/client/java/net/fabricmc/fabric/api/client/model/loading/v1/ModelResourceProvider.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.client.model.loading.v1;
+
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.client.render.model.UnbakedModel;
+import net.minecraft.util.Identifier;
+
+/**
+ * Interface for model resource providers.
+ *
+ * <p>Model resource providers hook the loading of model *files* from the resource tree;
+ * that is, in vanilla, it handles going from "minecraft:block/stone" to a
+ * "assets/minecraft/models/block/stone.json" file.
+ *
+ * <p>A common use of this is to cooperate with {@link ModelLoadingPlugin.Context#addModel} to directly
+ * add custom {@link UnbakedModel} instances.
+ *
+ * <p>This is also where you want to add your own custom model formats.
+ *
+ * <p>As each model reload uses a new provider, it is safe
+ * (and recommended!) to cache information inside a loader.
+ *
+ * <p>Keep in mind that only *one* ModelResourceProvider may respond to a given model
+ * at any time. If you're writing, say, an OBJ loader, this means you could
+ * easily conflict with another OBJ loader unless you take some precautions,
+ * for example:
+ *
+ * <ul><li>Only load files with a mod-suffixed name, such as .architect.obj,
+ * <li>Only load files from an explicit list of namespaces, registered elsewhere.</ul>
+ */
+@FunctionalInterface
+public interface ModelResourceProvider {
```

Comment:
Should both `ModelResourceProvider` and `ModelVariantProvider` be refactored to the same outer class? (`ModelProvider.Resource` and `ModelProvider.Variant`) This would align their design with the `ModelModifier`s.

Reply:
Done.